mbed_targets/_internal/target_attribute_hierarchy_parsers/accumulating_attribute_parser.py

```diff
@@ -0,0 +1,134 @@
+"""Functions for parsing the inheritance for accumulating attributes.
+
+Accumulating attributes are both defined and can be added to and removed from further down the hierarchy.
+The hierarchy is also slightly different to the other fields as it is determined as 'breadth-first' in
+multiple inheritance, so targets at a lower level will always take precedence over targets at a higher level.
+"""
+from collections import deque
+from typing import Dict, List, Any, Deque
+
+ACCUMULATING_ATTRIBUTES = ["extra_labels", "macros", "device_has", "features", "components"]
+ALL_ACCUMULATING_ATTRIBUTES = ACCUMULATING_ATTRIBUTES + [
+    "extra_labels_remove",
+    "extra_labels_add",
+    "macros_remove",
+    "macros_add",
+    "device_has_remove",
+    "device_has_add",
+    "features_remove",
+    "features_add",
+    "components_remove",
+    "components_add",
+]
+
+
+def _targets_accumulate_hierarchy(all_targets_data: Dict[str, Any], target_name: str) -> List[dict]:
+    """List all ancestors of a target in order of accumulation inheritance (breadth-first).
+
+    Using a breadth-first traverse of the inheritance tree, return a list of targets in the
+    order of inheritance, starting with the target itself and finishing with its highest ancestor.
+
+    Args:
+        all_targets_data: a dictionary representation of all the data in a targets.json file
+        target_name: the name of the target we want to calculate the attributes for
+
+    Returns:
+        A list of dicts representing each target in the hierarchy.
+    """
+    accumulation_order: List[dict] = []
+
+    still_to_visit: Deque[dict] = deque()
+    still_to_visit.appendleft(all_targets_data[target_name])
+
+    while still_to_visit:
```

Comment:
This and the hierarchy calculator for overriding attributes always seem to end up being pretty difficult to understand/read. I tried pulling bits out into separate functions but that seemed to make it even worse, though now Code Climate is unhappy about the complexity. Any suggestions?

Reply:
If there's any possibility of a recursive solution here that would probably be easier to grok. I'm still trying to figure out exactly what's going on here in my rusty brain so I don't have any concrete suggestions yet.

Reply:
There is at least for the overriding attributes hierarchy parser, but I have a bit of a hatred for recursive solutions, though I do understand that is probably just me. I could have a go at that and see if it ends up a bit easier to understand.

Reply:
Sometimes recursive solutions are terrible :) it just depends on the code, sometimes they can simplify code like this.

Reply:
Ok, I had a go at doing this recursively and I still ended up needing a queue in order to track previously traversed 'inherits' and it ended up more complicated than this (similar discussion here: https://cs.stackexchange.com/questions/64379/can-breadth-first-search-be-implemented-recursively-without-data-structures). The depth first search done in overriding attributes hierarchy can be done recursively, but part of me really likes the symmetry between the functions here.